price/pricing/network_load_calculator.go

```diff
@@ -0,0 +1,242 @@
+	conn, err := nlmc.db.Connection()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Release()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	rows, err := conn.Query(ctx, "SELECT multiplier, country_code FROM country_price_multipliers")
+	if err != nil {
+		return nil, err
+	}
+	results := make(map[ISO3166CountryCode]float64)
+	for rows.Next() {
+		var res = multiplierDB{}
+		if err := pgxscan.ScanRow(&res, rows); err != nil {
+			return nil, err
+		}
+		code := ISO3166CountryCode(res.CountryCode)
+		if code.Validate() != nil {
+			log.Error().Msgf("invalid country code %s in db", code)
+			continue
+		}
+		results[code] = res.Multiplier
+	}
+
+	return results, nil
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) updateCountries() {
+	log.Debug().Msgf("country update started")
+	newCache := make(map[ISO3166CountryCode]float64)
+	for k := range CountryCodeToName {
+		mul, err := nlmc.updateMultiplierForCountryByLoad(k)
+		if err != nil {
+			log.Err(err).Msgf("could not update multiplier for %s", k)
+		}
+		newCache[k] = mul
+		log.Trace().Msgf("updated multiplier for %s to %v", k, mul)
+	}
+	nlmc.setMultiplierMapCache(newCache)
+	log.Debug().Msgf("country update complete")
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) updateMultiplierForCountryByLoad(isoCode ISO3166CountryCode) (float64, error) {
+	mul, err := nlmc.fetchMultiplierFromProm(isoCode)
+	if err != nil {
+		return mul, err
+	}
+
+	return mul, nlmc.storeMultiplier(isoCode, mul)
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) storeMultiplier(isoCode ISO3166CountryCode, multiplier float64) error {
+	conn, err := nlmc.db.Connection()
+	if err != nil {
+		return err
+	}
+	defer conn.Release()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	query := `INSERT INTO country_price_multipliers(
+		country_code, multiplier, updated_at)
+		VALUES ($1, $2, now())
+		ON CONFLICT (country_code) DO UPDATE
+			SET multiplier = $2,
+				updated_at = now();`
+	_, err = conn.Exec(ctx, query, isoCode.String(), multiplier)
+	return err
+}
+
+func (nlmc *NetworkLoadMultiplierCalculator) fetchMultiplierFromProm(isoCode ISO3166CountryCode) (float64, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*15)
+	defer cancel()
+	providers, err := nlmc.getActiveProvidersForCountry(ctx, isoCode)
+	if err != nil {
+		return 0, err
+	}
+
+	activeSessions, err := nlmc.getSessionsForCountry(ctx, isoCode)
+	if err != nil {
+		return 0, err
+	}
+
+	mul := nlmc.calculateMultiplier(providers, activeSessions)
+	return mul, nil
+}
+
+// calculateMultiplier calculates a coefficient for price multiplication.
+// It returns a value of 0.5 < x < 2.0.
+// If there are more active sessions than providers, we want to incentivise providers by increasing service price in that region.
+func (nlmc *NetworkLoadMultiplierCalculator) calculateMultiplier(providers, activeSessions int64) float64 {
```

Comment:
This one is easy to test and would be nice to test so that someone doesn't change it by accident and start diving by zero or screw up the values all together.